(How) does weather influence ammount of bikeleases?

In [21]:
#Data Analysis libraries
import numpy as np
import pandas as pd


In [22]:
#importing needed csv files
stations = pd.read_csv('input/station.csv')
trips = pd.read_csv('input/trip.csv', error_bad_lines=False)
weathers = pd.read_csv('input/weather.csv')

b'Skipping line 50794: expected 12 fields, saw 20\n'


In [23]:
#convert the strings to datetime objects
trips['starttime'] = pd.to_datetime(trips['starttime'])
trips['stoptime'] = pd.to_datetime(trips['stoptime'])


In [24]:
#trim the time of, we only want the date
trips['starttime'] = trips['starttime'].dt.date
trips['stoptime'] = trips['stoptime'].dt.date

In [25]:
#group by trips per day, add trips per day to weather dataframe
tripsperday = trips.groupby(['starttime']).size().reset_index(name='biketrips')
weathers['biketrips'] = tripsperday['biketrips']
weathers

Date  Max_Temperature_F  Mean_Temperature_F  Min_TemperatureF  \
0    10/13/2014                 71                62.0                54   
1    10/14/2014                 63                59.0                55   
2    10/15/2014                 62                58.0                54   
3    10/16/2014                 71                61.0                52   
4    10/17/2014                 64                60.0                57   
5    10/18/2014                 68                64.0                59   
6    10/19/2014                 73                64.0                55   
7    10/20/2014                 66                60.0                55   
8    10/21/2014                 64                58.0                55   
9    10/22/2014                 60                58.0                57   
10   10/23/2014                 62                55.0                50   
11   10/24/2014                 60                56.0                51   
12   10/25/2014                 64                58.0                52   
13   10/26/2014                 59                52.0                48   
14   10/27/2014                 62                54.0                45   
15   10/28/2014                 62                56.0                51   
16   10/29/2014                 64                59.0                54   
17   10/30/2014                 62                58.0                55   
18   10/31/2014                 59                54.0                52   
19    11/1/2014                 55                52.0                48   
20    11/2/2014                 57                51.0                45   
21    11/3/2014                 60                58.0                55   
22    11/4/2014                 61                58.0                55   
23    11/5/2014                 61                57.0                53   
24    11/6/2014                 64                60.0                54   
25    11/7/2014                 59                52.0                45   
26    11/8/2014                 55                48.0                42   
27    11/9/2014                 57                52.0                48   
28   11/10/2014                 54                48.0                43   
29   11/11/2014                 48                42.0                36   
..          ...                ...                 ...               ...   
659    8/2/2016                 73                64.0                57   
660    8/3/2016                 75                66.0                60   
661    8/4/2016                 80                69.0                59   
662    8/5/2016                 79                70.0                61   
663    8/6/2016                 72                63.0                55   
664    8/7/2016                 72                66.0                60   
665    8/8/2016                 73                64.0                57   
666    8/9/2016                 72                66.0                60   
667   8/10/2016                 75                68.0                60   
668   8/11/2016                 80                70.0                61   
669   8/12/2016                 87                75.0                63   
670   8/13/2016                 90                78.0                66   
671   8/14/2016                 82                71.0                60   
672   8/15/2016                 84                72.0                61   
673   8/16/2016                 81                69.0                57   
674   8/17/2016                 79                70.0                60   
675   8/18/2016                 86                73.0                60   
676   8/19/2016                 95                60.0                26   
677   8/20/2016                 91                78.0                64   
678   8/21/2016                 73                66.0                60   
679   8/22/2016                 72                64.0     

In [26]:
#sum the ammount of trips, grouped per unique event
Eventtrips = weathers.groupby('Events').apply(lambda x: x.groupby('Date').biketrips.first().sum()).reset_index(name='tripsperevent')

In [27]:
#refactor the information into a new dataframe to get an overview of how many days each event took place
bla = weathers['Events'].value_counts().reset_index(name='count')
bla = bla.sort_values(by=['index'])
bla.rename(columns={'index': 'Events'}, inplace=True)
bla

Events  count
1                  Fog     16
2           Fog , Rain      7
3             Fog-Rain      6
0                 Rain    287
8          Rain , Snow      1
5  Rain , Thunderstorm      3
6            Rain-Snow      2
4    Rain-Thunderstorm      4
7                 Snow      2

In [28]:
#Calculate average trips per event per day
Eventtrips['count'] = bla['count']
Eventtrips
Merge = pd.merge(Eventtrips, bla, on='Events'  )
Merge = Merge[['Events','tripsperevent', 'count_y']]
Merge['avg'] = Merge['tripsperevent']/Merge['count_y']
Merge

Events  tripsperevent  count_y         avg
0                  Fog           6528       16  408.000000
1           Fog , Rain           3302        7  471.714286
2             Fog-Rain           1535        6  255.833333
3                 Rain         101934      287  355.170732
4          Rain , Snow            126        1  126.000000
5  Rain , Thunderstorm           1355        3  451.666667
6            Rain-Snow            252        2  126.000000
7    Rain-Thunderstorm           1032        4  258.000000
8                 Snow            463        2  231.500000

In [29]:
#Calculate trips per day where Event is null (good weather), and add to dataframe
df = pd.DataFrame([['NaN',weathers['biketrips'].sum(),weathers['Events'].isnull().sum(),weathers['biketrips'].sum()/weathers['Events'].isnull().sum()]], columns=('Events','tripsperevent','count_y','avg'))
Merge = Merge.append(df, ignore_index=True)
Merge


Events  tripsperevent  count_y         avg
0                  Fog           6528       16  408.000000
1           Fog , Rain           3302        7  471.714286
2             Fog-Rain           1535        6  255.833333
3                 Rain         101934      287  355.170732
4          Rain , Snow            126        1  126.000000
5  Rain , Thunderstorm           1355        3  451.666667
6            Rain-Snow            252        2  126.000000
7    Rain-Thunderstorm           1032        4  258.000000
8                 Snow            463        2  231.500000
9                  NaN         286857      361  794.617729

conclusion: the weather certainly influences the ammount of leased bicycles. Rain and snow is the the combination that effects the lease the most (126 leases on average). Now that we have seen the absolute values, we want to put this into a nice graph and calculate the correlation

In [31]:
#importing needed csv files again, to reset all changes
station = pd.read_csv('input/station.csv')
trip = pd.read_csv('input/trip.csv', error_bad_lines=False)
weather = pd.read_csv('input/weather.csv')

b'Skipping line 50794: expected 12 fields, saw 20\n'


In [32]:
#refactor columns to needed date and datetime
weather['Date'] = pd.to_datetime(weather['Date'], format="%m/%d/%Y")
weather['year'] = weather['Date'].dt.year
weather['month'] = weather['Date'].dt.month

In [33]:
#calc temp and precipation per month, year. Concatenate them into 1 dataframe
temppermy = weather[['Mean_Temperature_F', 'year', 'month']].groupby(['year','month']).mean()
precpermy = weather[['Precipitation_In', 'year', 'month']].groupby(['year','month']).mean()
result = pd.concat([tripspermy, temppermy, precpermy], axis=1, join='inner')


NameError: name 'tripspermy' is not defined

- Deel door neerslag omdat je verwacht dat het aantal trips groter is bij lage neerslag
- Vermenigvuldig met de temperatuur omdat je verwacht dat het aantal trips grote is bij een hogere temperatuur

In [39]:
#amount of trips / precipitation
result['TripsPrec'] = tripspermy['Amount of trips']/precpermy['Precipitation_In']
#amount of trips / temperature
result["TripsTemp"] = tripspermy['Amount of trips']*temppermy['Mean_Temperature_F']
result

NameError: name 'tripspermy' is not defined

Correlatie tussen neerslag en fietsverhuur (line-graph)

X as = het aantal waarnemingen (de maanden)
Y as = trips/neerslag

In [35]:
result['TripsPrec'].plot()
plt.xlabel('Tijd')
plt.ylabel('Neerslag')

NameError: name 'result' is not defined

Met neerslag is dus wel significant anders, er worden duidelijk minder trips genomen op het moment dat het regent. Maar neerslag is niet de enige definitie van slecht weer. Temperatuur speelt hier ook een grote rol in. We gaan kijken of dit ook invloed heeft en wat die invloed is.

Correlatie tussen temperatuur en fietsverhuur (line-graph)

X as = het aantal waarnemingen (de maanden)
Y as = trips*temperatuur



In [40]:
result[['TripsTemp']].plot()
plt.xlabel('Tijd')
plt.ylabel('Temperatuur')

NameError: name 'result' is not defined

Er is hier dus geen verschil, is namelijk bijna hetzelfde als amount of trips over time (zoals hierboven te zien)



We kunnen dus concluderen dat de neerslag (maar ook het type neerslag) significante invloed heeft op het aantal gehuurde fietsen. Temperatuur daarentegen heeft geen significante invloed. Van alle typen neerslag heeft de combinatie van sneeuw en regen de meeste invloed op het aantal huurders op die dag. 